persistence/sql/persister_consent.go

```diff
@@ -439,84 +439,103 @@ func (p *Persister) VerifyAndInvalidateLogoutRequest(ctx context.Context, verifi
-			notAfter).Exec()
+	challenges := []string{}
+
+	// Select challenges from all consent requests that can be safely deleted with limit
+	// where hydra_oauth2_consent_request were unhandled or rejected, so either of these is true
+	// - hydra_oauth2_authentication_request_handled does not exist (unhandled)
+	// - hydra_oauth2_consent_request_handled has valid error (rejected)
+	// AND timed-out
+	// - hydra_oauth2_consent_request.requested_at < minimum between ttl.login_consent_request and notAfter
+	q := p.Connection(ctx).RawQuery(fmt.Sprintf(`
+		SELECT %[1]s.challenge
+		FROM %[1]s
+		LEFT JOIN %[2]s ON %[1]s.challenge = %[2]s.challenge
+		WHERE (
+			(%[2]s.challenge IS NULL)
+			OR (%[2]s.error IS NOT NULL AND %[2]s.error <> '{}' AND %[2]s.error <> '')
+		)
+		AND %[1]s.requested_at < ?
+		`,
+		(&cr).TableName(),
+		(&crh).TableName()),
+		notAfter)
+
+	if err := q.Limit(limit).All(&challenges); err == sql.ErrNoRows {
+		return errors.Wrap(fosite.ErrNotFound, "")
+	}
 
-		if err != nil {
+	// Delete in batch consent requests and their references in cascade
+	for i := 0; i < len(challenges); i += batchSize {
+		j := i + batchSize
+		if j > len(challenges) {
+			j = len(challenges)
+		}
+
+		q := p.Connection(ctx).RawQuery(
+			fmt.Sprintf("DELETE FROM %s WHERE challenge in (?)", (&cr).TableName()),
+			challenges[i:j],
+		)
+
+		if err := q.Exec(); err != nil {
 			return sqlcon.HandleError(err)
 		}
+	}
 
-		// This query is needed due to the fact that the first query will not delete cascade to the consent tables
-		// This cleans up the consent requests if requests have timed out or been rejected.
-
-		// Using NOT EXISTS instead of LEFT JOIN or NOT IN due to
-		// LEFT JOIN not supported by Postgres and NOT IN will have performance hits with large tables.
-		// https://stackoverflow.com/questions/19363481/select-rows-which-are-not-present-in-other-table/19364694#19364694
-		// Cannot use table aliasing in MYSQL, will work in Postgresql though...
-		err = p.Connection(ctx).RawQuery(
-			fmt.Sprintf(`
-			DELETE
-			FROM %[1]s
-			WHERE NOT EXISTS
-				(
-				SELECT NULL
-				FROM %[2]s
-				WHERE %[1]s.challenge = %[2]s.challenge AND (%[2]s.error = '{}' OR %[2]s.error = '' OR %[2]s.error is NULL)
-				)
-			AND requested_at < ?
-			AND requested_at < ?`,
-				(&cr).TableName(),
-				(&crh).TableName()),
-			time.Now().Add(-p.config.ConsentRequestMaxAge()),
-			notAfter).Exec()
+	// Select challenges from all authentication requests that can be safely deleted with limit
+	// where hydra_oauth2_authentication_request were unhandled or rejected, so either of these is true
+	// - hydra_oauth2_authentication_request_handled does not exist (unhandled)
+	// - hydra_oauth2_authentication_request_handled has valid error (rejected)
+	// AND timed-out
+	// - hydra_oauth2_authentication_request.requested_at < minimum between ttl.login_consent_request and notAfter
+	q = p.Connection(ctx).RawQuery(fmt.Sprintf(`
+		SELECT %[1]s.challenge
```

Comment:
Looks like the queries are identical - could we use one const and printf to generate them instead?

Reply:
Yeah, I think I can give that a try!